File: app/src/main/java/gargoyle/rpycg/service/LocaleConverter.java
package gargoyle.rpycg.service;

import gargoyle.fx.FXContext;
import gargoyle.fx.FXContextFactory;
import gargoyle.fx.FXUtil;
import gargoyle.fx.log.FXLog;
import gargoyle.rpycg.ex.AppUserException;

import java.text.MessageFormat;
import java.util.Arrays;
import java.util.Collection;
import java.util.Locale;
import java.util.Map;
import java.util.MissingResourceException;
import java.util.Objects;
import java.util.Optional;
import java.util.ResourceBundle;
import java.util.Set;
import java.util.stream.Collectors;

public final class LocaleConverter {
    private static final String KEY_SUPPORTED_LOCALES = "supported_locales";
    private final ResourceBundle resources;

    public LocaleConverter(final FXContext context) {
        this(context.getBaseClass(), FXContextFactory.currentContext().getLocale());
    }

    LocaleConverter(final Class<?> baseClass, final Locale locale) {
        resources = Optional.of((LocaleConverter.class))
                .map(FXUtil::resolveBaseName)
                .map(baseName -> {
                    try {
                        return (ResourceBundle.getBundle(baseName, locale, baseClass.getClassLoader()));
                    } catch (final MissingResourceException e) {
                        FXLog.error("Can''t find resources for {baseName}", Map.of("baseName", baseName));
                        return null;
                    }
                }).orElseThrow(() ->
                        new AppUserException("No resources {resource}",
                                Map.of("resource", LocaleConverter.class.getName())));
    }

    public Set<Locale> getLocales() {
        return Arrays.stream(resources.getString(KEY_SUPPORTED_LOCALES).split(","))
                .map(this::toLocale).collect(Collectors.toSet());
    }

    public Locale getSimilarLocale(final Locale locale) {
        return getSimilarLocale(getLocales(), locale);
    }

    public String toDisplayString(final Locale locale) {
        return MessageFormat.format("{0} ({1})",
                locale.getDisplayLanguage(FXContextFactory.currentContext().getLocale()),
                locale.getDisplayLanguage(locale));
    }

    public Locale toLocale(final String loc) {
        final String[] parts = loc.trim().split("_");
        return switch (parts.length) {
            case 3 -> new Locale(parts[0], parts[1], parts[2]);
            case 2 -> new Locale(parts[0], parts[1]);
            case 1 -> new Locale(parts[0]);
            case 0 -> new Locale(loc);
            default -> new Locale(loc);
        };
    }

    public String toString(final Locale locale) {
        final boolean hasLanguage = !locale.getLanguage().isEmpty();
        final boolean hasScript = !locale.getScript().isEmpty();
        final boolean hasCountry = !locale.getCountry().isEmpty();
        final boolean hasVariant = !locale.getVariant().isEmpty();
        final StringBuilder result = new StringBuilder(locale.getLanguage());
        if (hasCountry || hasLanguage && (hasVariant || hasScript)) {
            result.append('_').append(locale.getCountry());
        }
        if (hasVariant && (hasLanguage || hasCountry)) {
            result.append('_').append(locale.getVariant());
        }
        if (hasScript && (hasLanguage || hasCountry)) {
            result.append("_#").append(locale.getScript());
        }
        return result.toString();
    }

    private Locale getSimilarLocale(final Collection<Locale> locales, final Locale locale) {
        return locales.stream().filter(loc -> Objects.equals(locale.getLanguage(), loc.getLanguage())).findAny()
                .orElseGet(() -> locales.stream().findAny().orElse(Locale.getDefault()));
    }
}

File: app/src/main/kotlin/gargoyle/rpycg/service/LocaleConverter.kt
package gargoyle.rpycg.service

import gargoyle.fx.FxContext
import gargoyle.fx.FxUtil.get
import java.util.Locale
import java.util.ResourceBundle

class LocaleConverter internal constructor(context: FxContext) {

    private val resources: ResourceBundle =
        context.loadResources(LocaleConverter::class) ?: error("no resources {LocaleConverter}")

    val locales: Set<Locale>
        get() = resources[KEY_SUPPORTED_LOCALES].split(",").map { toLocale(it) }.toSet()

    fun getSimilarLocale(locale: Locale): Locale =
        locales.firstOrNull { locale.language == it.language } ?: locales.firstOrNull() ?: Locale.getDefault()

    fun toDisplayString(locale: Locale): String =
        "${locale.getDisplayLanguage(FxContext.current.locale)} (${locale.getDisplayLanguage(locale)})"

    fun toLocale(loc: String): Locale {
        val parts = loc.trim().split("_")
        return when (parts.size) {
            3 -> Locale(parts[0], parts[1], parts[2])
            2 -> Locale(parts[0], parts[1])
            1 -> Locale(parts[0])
            0 -> Locale(loc)
            else -> Locale(loc)
        }
    }

    fun toString(locale: Locale): String =
        with(mutableListOf<String>(locale.language)) {
            if (locale.country.isNotEmpty() || locale.language.isNotEmpty() && (locale.variant.isNotEmpty() || locale.script.isNotEmpty()))
                this += locale.country
            if (locale.variant.isNotEmpty() && (locale.language.isNotEmpty() || locale.country.isNotEmpty()))
                this += locale.variant
            if (locale.script.isNotEmpty() && (locale.language.isNotEmpty() || locale.country.isNotEmpty()))
                this += "#${locale.script}"
            this
        }.joinToString("_")

    companion object {
        private const val KEY_SUPPORTED_LOCALES = "supported_locales"
    }
}
